### PR TITLE
Amazon S3 Data Source and Google Drive Data Source support

### PR DESCRIPTION
**Pull request by hidekoji**
# Description

Added below APIs for Amazon S3 and Google Drive Data Source Support.

- clearGoogleDriveCacheFile
- clearS3CacheFile
- downloadDataFileFromS3  
- downloadDataFileFromGoogleDrive
- getCSVFileFromS3
- getCSVFilesFromS3
- getExcelFileFromS3
- getExcelFilesFromS3
- getExcelSheetsFromGoogleDriveExcelFile
- getExcelSheetsFromS3ExcelFile
- guessFileEncodingForS3File
- guessFIleEcondingForGoogleDriveFile
- getGoogleTokenForDrive
- getCSVFileFromGoogleDrive
- getCSVFilesFromGoogleDrive
- getExcelFileFromGoogleDrive
- getExcelFilesFromGoogleDrive

# Checklist
Make sure you have performed following items before submitting this pull request.
If not, please describe the reason.  

- [ ] Add test cases for this fix/enhancement
- [ ] Pass devtools::check()
- [ ] Pass devtools::test()
- [ ] Test installing from github
- [x] Tested with Exploratory

### CODE REVIEW

File: R/aws_s3.R [hunk 1]
```diff
@@ -0,0 +1,148 @@
+#' API to download remote data file (excel, csv) from Amazon S3 and cache it if necessary
+#' it uses tempfile https://stat.ethz.ch/R-manual/R-devel/library/base/html/tempfile.html
+#' and a R variable with name of hashed region, bucket, key, secret, fileName are  assigned to the path given by tempfile.
+downloadDataFileFromS3 <- function(region, bucket, key, secret, fileName, as = "text"){
+  shouldCacheFile <- getOption("tam.should.cache.datafile")
+  filepath <- NULL
+  hash <- digest::digest(stringr::str_c(region, bucket, fileName, sep = ":"), "md5", serialize = FALSE)
+  tryCatch({
+    filepath <- eval(as.name(hash))
```
Comment (hideaki) [inline R/aws_s3.R:9]: Check what this was.

File: R/aws_s3.R [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    #
+    # The following command is harmless even if you have the directory already.
+    # http://stackoverflow.com/questions/4216753/check-existence-of-directory-and-create-if-doesnt-exist
+    dir.create(tempdir(), showWarnings = FALSE)
+
+    # download file to temporary location
+    aws.s3::save_object(fileName, bucket = bucket, as = as, region = region, key = key, secret = secret, file = tmp)
+    # cache file
+    if(!is.null(shouldCacheFile) && isTRUE(shouldCacheFile)){
+      assign(hash, tmp, envir = .GlobalEnv)
```
Comment (hideaki) [inline R/aws_s3.R:39]: Check what this was.

File: R/google_drive.R [hunk 3]
```diff
@@ -0,0 +1,189 @@
+#' API to get a list of files in Google Drive
+#' @export
+listItemsInGoogleDrive <- function(teamDriveId = "", path = NULL, type =  c("csv", "tsv", "txt", "folder", "xls", "xlsx")){
+  if(!requireNamespace("googledrive")){stop("package googledrive must be installed.")}
+  token = getGoogleTokenForDrive()
+  googledrive::drive_set_token(token)
+  if (!is.null(path)) {
```
Comment (hideaki) [inline R/google_drive.R:7]: Add comment on our decision on ID vs path.

File: R/google_drive.R [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    path = googledrive::as_id(path)
+  }
+  #if team id is provided search documents within the team.
```
Comment (hideaki) [inline R/google_drive.R:10]: Correct comment format.

File: R/google_drive.R [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+  if(teamDriveId != "" && !is.null(teamDriveId)) {
```
Comment (hideaki) [inline R/google_drive.R:11]: Adjust format for if statement.

File: R/google_drive.R [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    # To improve performance, only get id, name, mimeType, modifiedTime,  size, parents for each file.
```
Comment (hideaki) [inline R/google_drive.R:12]: Improve format in the comment.

File: R/google_drive.R [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    googledrive::drive_ls(path = path, type = type, team_drive=googledrive::as_id(teamDriveId) ,pageSize=1000, fields="files/id, files/name, files/mimeType, files/modifiedTime, files/size, files/parents, nextPageToken")
```
Comment (hideaki) [inline R/google_drive.R:13]: Space after comma.

File: R/google_drive.R (hunk 5 above)
Comment (hideaki) [inline R/google_drive.R:11]: Check if we can pass NULL teamDriveId and consolidate code for those 2 cases.

File: R/google_sheets.R [hunk 8]
```diff
@@ -119,9 +119,14 @@ getGoogleSheetWorkSheetList <- function(tokenFileId = "", title, id = NULL){
   googlesheets4::sheets_sheets(sheet)
 }
 
-getTeamDrives <- function(tokenFileId = ""){
+getTeamDrives <- function(tokenFileId = "", useGoogleSheetToken = TRUE){
```
Comment (hideaki) [inline R/google_sheets.R:122]: Add comment on useGoogleSheetsToken.

File: R/oauth.R [hunk 9]
```diff
@@ -122,13 +122,46 @@ getGoogleTokenForSheet <- function(tokenFileId="", useCache=TRUE){
...
-
 #' API to refresh token
 #' @export
 refreshGoogleTokenForSheet <- function(tokenFileId){
   getGoogleTokenForSheet(tokenFileId, FALSE)
 }
 
+#' @export
+getGoogleTokenForDrive <- function(tokenFileId = "", useCache=TRUE){
+  # As per Kan, this can be hard coded since Google limits acces per ViewID (tableID) and
```
Comment (hideaki) [inline R/oauth.R:133]: This comment look irrelevant.

File: R/oauth.R [hunk 10]
```diff
@@ -122,13 +122,46 @@ getGoogleTokenForSheet <- function(tokenFileId="", useCache=TRUE){
...
+  # not by clientID
+  appName = "google"
+  # retrieve token info from environment
+  # main purpose is to enable server refresh
+  token_info <- getTokenInfo("googledrive") # this should be googlesheets but our plunin is already named googledrive
```
Comment (hideaki) [inline R/oauth.R:138]: Typo: plunin

File: R/google_drive.R (hunk 3 above)
Comment (hidekoji) [reply R/google_drive.R:10]: Added comment.

File: R/google_drive.R (hunk 4 above)
Comment (hidekoji) [reply R/google_drive.R]: fixed the format.

File: R/google_drive.R (hunk 5 above)
Comment (hidekoji) [reply R/google_drive.R]: fixed the if statement format.

File: R/google_drive.R (hunk 6 above)
Comment (hidekoji) [reply R/google_drive.R]: fixed the format.

File: R/google_drive.R (hunk 7 above)
Comment (hidekoji) [reply R/google_drive.R]: fixed the spacing.

File: R/google_drive.R (hunk 5 above)
Comment (hidekoji) [reply R/google_drive.R]: Changed to pass NULL as suggested.

File: R/google_sheets.R (hunk 8 above)
Comment (hidekoji) [reply R/google_sheets.R]: Added comment.

File: R/oauth.R (hunk 9 above)
Comment (hidekoji) [reply R/oauth.R]: Deleted unnecessary comment.

File: R/oauth.R (hunk 10 above)
Comment (hidekoji) [reply R/oauth.R]: Removed irrelevant comment.

File: R/google_drive.R [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+  if (!is.null(path)) {
+    path = googledrive::as_id(path)
+  }
+  df <- NULL
+  #if team id is provided search documents within the team.
+  # If team id is provided search documents within the team.
+  if (teamDriveId != "" && !is.null(teamDriveId)) {
+    teamDriveId = googledrive::as_id(teamDriveId)
+  }
+  df <- googledrive::drive_get(team_drive = googledrive::as_id(teamDriveId), id = path)
```
Comment (hideaki) [inline R/google_drive.R:40]: Since you already called as_id at line 38, I think you don't need as_id here again. Please test after this change.

Comment (hidekoji) [reply R/google_drive.R:40]: removed the duplicated as_id call.
